### ISSUE TITLE
Closing a tab not warning of unsaved changes

### ISSUE THREAD

**Issue by sunsetsystems** [openemr/openemr#290]
Using Tabs layout, I tested making a change in a patient's demographics and then closing the tab without saving the change.  The tab closed and there was no warning.  Also tested the same thing in the layout editor page (Administration -> Layouts) with the same result.  Did not test other pages but I suspect it's a general problem with all.

Also regarding closing tabs, I see that the page is left empty and no existing tab is shown in its place.  This seems wrong.

**Comment by bradymiller** [comment]
My thoughts here, are that in the tabs layout, need the following to happen when a user is editing a form:
1. Warn of unsaved changes if try to close the tab.
2. Warn of unsaved changes if try to open another screen that will overwrite the tab with the form in it.

**Comment by sunsetsystems** [comment]
Agree.  But seems the beforeunload handler should trigger in either case?  So maybe mostly a matter of finding out why it's not working.

**Comment by rishabhemr** [comment]
We have fixed this issue along with the issue #260. Please review the code and let us know the feedback.

Rishabh Software

**Comment by bradymiller** [comment]
Hi,
Can you post the code as a Pull Request?
thanks,
-brady

**Comment by bradymiller** [comment]
Reviews this fix. Rather than ask if ok to close the tab (which will become tedious for users), need to incorporate the beforeunload handler in the pertinent forms. To see example, recommend looking at use of this in the openemr/interface/forms/LBF/new.php script.

**Comment by bradymiller** [comment]
@sunsetsystems ,
I am only seeing the  beforeunload handler being used in the openemr/interface/forms/LBF/new.php script. Were there other places that had a mechanism to prevent losing information in the frames UI?

**Comment by sunsetsystems** [comment]
Yes it was also done for demographics and the layout editor.  Commit is here:
https://github.com/openemr/openemr/commit/364afe27cfa3621f93eaee07834b4821d0827c66

**Comment by rishabhemr** [comment]
We are working on this and will update the status next week.

**Comment by rishabhemr** [comment]
We have made the changes in the above commit regarding closing tab with drag and drop tabs feature.
#251 and #290

Rishabh Software

**Comment by robertdown** [comment]
Hi @rishabhemr can you submit a pull request here on github with those changes for review? Thanks!

**Comment by tywrenn** [comment]
Bump -- this should've been added

**Comment by stephenwaite** [comment]
we're getting closer with https://github.com/openemr/openemr/pull/3303

**Comment by tywrenn** [comment]
@stephenwaite Looks like that merged -- time to close this

**Comment by stephenwaite** [comment]
hi @tywrenn , that was just for the soap form
https://github.com/openemr/openemr/pull/3303#issuecomment-611363816

**Comment by tywrenn** [comment]
Bump

**Comment by kojiromike** [comment]
Trying something out here:

@copilot is this still a valid issue today?

**Comment by kojiromike** [comment]
## Analysis by Claude (AI Assistant)

I investigated whether this issue is still valid against the current master branch.

### Findings:

**Part 1: No warning when closing tabs with unsaved changes** - **Still an issue**

While `beforeunload` handlers have been added to the specific pages mentioned:
- `interface/patient_file/summary/demographics_full.php` (lines 206-212)
- `interface/super/edit_layout.php` (lines 2206-2212)

These handlers only trigger on browser-level navigation (closing browser tab, navigating away). They do **not** work when closing a tab via OpenEMR's internal tab system.

The `tabClose()` function in `interface/main/tabs/js/tabs_view_model.js:119-127` simply removes the tab with no check for unsaved changes:

```javascript
function tabClose(data,evt)
{
    //remove the tab
    app_view_model.application_data.tabs.tabsList.remove(data);
    //activate the next tab
    if(data.visible()) {
        activateTab(app_view_model.application_data.tabs.tabsList()[app_view_model.application_data.tabs.tabsList().length-1]);
    }
}
```

**Part 2: Empty page after closing tab** - **Appears to be fixed**

The `tabClose()` function now calls `activateTab()` on another tab after closing.

### What's still needed:

1. A mechanism for iframe content to communicate dirty state to the parent tab system
2. Logic in `tabClose()` to check for unsaved changes before closing
3. A global dirty state tracking system that integrates with the tabs framework

This issue remains valid for the unsaved changes warning functionality.
